### ISSUE TITLE
CRW set DateTimeOriginal not working correctly on MinGW 

### ISSUE THREAD

**Issue by clanmills** [Exiv2/exiv2#1219]
**On any platform _(except MinGW)_:**

```
...test $ cp data/exiv2-canon-powershot-s40.crw test.crw
...test $ exiv2 -g Original test.crw 
Exif.Photo.DateTimeOriginal                  Ascii      20  2005:04:23 17:54:36
...test $ exiv2 -M'set Exif.Photo.DateTimeOriginal 2020:05:25 13:39:00' test.crw
...test $ exiv2 -g Original test.crw 
Exif.Photo.DateTimeOriginal                  Ascii      20  2020:05:25 13:39:00
$ 
```

**On MinGW:**

```
...test $ cp data/exiv2-canon-powershot-s40.crw test.crw
...test $ exiv2 -g Original test.crw 
Exif.Photo.DateTimeOriginal                  Ascii      20  2005:04:23 17:54:36
...test $ exiv2 -M'set Exif.Photo.DateTimeOriginal 2020:05:25 13:39:00' test.crw
...test $ exiv2 -g Original test.crw 
Exif.Photo.DateTimeOriginal                  Ascii      20  2106:02:07 06:28:15
$
```

**Discussion**

I've discovered this with the test harness on the final build before Exiv2 v0.27.3 RC2.

I cannot fix this for v0.27.3 because I don't know anything about the CRW format, so this requires considerable work to understand and fix.

**Workaround**

I'm going to modify test/crw-test.sh to omit this part of the test.  When this gets fixed in src/crwimage.cpp, we should restore the test.

**Comment by clanmills** [comment]
It's not odd that this has surfaced at this time.  test/crw-test.sh has not been active in the test harness for some time.  I only discovered last week that it wasn't being used.

It's odd that this only happens on MinGW as I believe dates are strings and have no interaction with platform time functions such as `localtime()`.

The issue is coming from src/crwimage_int.cpp.  Exif.Photo.DateTimeOriginal is tag 0x9003 (36867)
```
520 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance $ taglist | grep Exif.Photo.DateTimeOriginal
DateTimeOriginal,	36867,	0x9003,	Photo,	Exif.Photo.DateTimeOriginal,	Ascii,	The date and time when the original image data was generated. For a digital still camera the date and time the picture was taken are recorded.
521 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance $ calc "0x9003"
36867
522 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance $ 
```
The following code in crwimage_int.cpp sets up the encoder:
```
const CrwMapping CrwMap::crwMapping_[] = {
...
        CrwMapping(0x180e, 0x300a,   0, 0x9003, exifId,  decode0x180e, encode0x180e),
...
}; // CrwMap::crwMapping_[]
```
Here's the troublesome code:
```
void CrwMap::encode0x180e(const Image&      image,
                          const CrwMapping* pCrwMapping,
                                CiffHeader* pHead)
{
    assert(pCrwMapping != 0);
    assert(pHead != 0);

    time_t t = 0;
    const ExifKey key(pCrwMapping->tag_, Internal::groupName(pCrwMapping->ifdId_));
    const ExifData::const_iterator ed = image.exifData().findKey(key);
    if (ed != image.exifData().end()) {
        struct tm tm;
        std::memset(&tm, 0x0, sizeof(tm));
        int rc = exifTime(ed->toString().c_str(), &tm);
        if (rc == 0) t = timegm(&tm);
    }
    if (t != 0) {
        DataBuf buf(12);
        std::memset(buf.pData_, 0x0, 12);
        ul2Data(buf.pData_, static_cast<uint32_t>(t), pHead->byteOrder());
        pHead->add(pCrwMapping->crwTagId_, pCrwMapping->crwDir_, buf);
    }
    else {
        pHead->remove(pCrwMapping->crwTagId_, pCrwMapping->crwDir_);
    }
} // CrwMap::encode0x180e
```

It looks to me that he's calculating a uint32_t.  That's probably the offset in the file to the 20 byte string.   I'll step in the debugger.  Probably use QtCreator on MinGW.

**Comment by clanmills** [comment]
This function does interact with platform time functions.  exifTime() parses the string into a `struct tm tm`.  I'll figure this out.

**Comment by clanmills** [comment]
I know what's wrong.  It's the symbol EXV_HAVE_GMTIME_R which is only defined on MinGW and only used in two locations - both in src/crwimage_int.cpp.  So that explains why this has only surfaced on MinGW.

And I know why I haven't seen it before.  I changed the build tools on MinGW last week.  I am building again with GCC 7.4.0.  Last week, I was using 9.3.0.

I made the changes and to my surprise, the issue isn't fixed.  More debugging on Tuesday.  I'll test all DateTime tags, and restore test/crw-test.sh.

I feel it will safe to submit this fix when it's ready.  We have isolated the cause and know this only applies to crw files on MinGW.  So the fix is destined for RC2 (schedule 2020:05:31).
